marlowe-playground-client/src/Blockly/Dom.purs

```diff
@@ -0,0 +1,253 @@
+      { left, right } <- separate <$> traverse readAsBlockChild elementChildren
+      let
+        children = Object.fromFoldable left
+      pure
+        $ Tuple
+            (Block { id: blockId, type: blockType, children })
+            (join right)
+
+  -- Tries to read an Element as the children of a Block node. The Left hand side of the return represents a Block direct children
+  -- while the Right hand side represents the parsed `<next>` elements which should be appended to a parents Statement.
+  readAsBlockChild ::
+    forall m.
+    MonadEffect m =>
+    MonadThrow ReadDomError m =>
+    Element ->
+    m (Either ChildWithName (Array Block))
+  readAsBlockChild element = do
+    case Element.tagName element of
+      "field" -> do
+        name <- getAttribute "name" element
+        value <- getElementText element
+        pure $ Left $ Tuple name $ Field value
+      "statement" -> do
+        name <- getAttribute "name" element
+        child <- getSingleChild element
+        Tuple block nexts <- readAsBlock child
+        pure $ Left $ Tuple name $ Statement ([ block ] <> nexts)
+      "value" -> do
+        name <- getAttribute "name" element
+        child <- getSingleChild element
+        Tuple block nexts <- readAsBlock child
+        -- Even if a value can have a Block appended, it should have a single block child and no "siblings"
+        if length nexts /= 0 then
+          throwError $ IncorrectSiblingNesting "value"
+        else
+          pure $ Left $ Tuple name $ Value block
+      "next" -> do
+        child <- getSingleChild element
+        Tuple block nexts <- readAsBlock child
+        pure $ Right ([ block ] <> nexts)
+      _ -> throwError $ TypeMismatch element "field, statement, value, next"
+
+  getSingleChild :: forall m. MonadEffect m => MonadThrow ReadDomError m => Element -> m Element
+  getSingleChild element = do
+    children <- getChildrens element
+    if length children /= 1 then
+      throwError $ SingleChildExpected element $ length children
+    else do
+      pure $ unsafePartial $ UnsafeArray.head children
+
+  getChildrens :: forall m. MonadEffect m => Element -> m (Array Element)
+  getChildrens element =
+    liftEffect do
+      elements <- (ParentNode.children <<< Element.toParentNode) element
+      HTMLCollection.toArray elements
+
+  getChildNodes :: forall m. MonadEffect m => Element -> m (Array Node)
+  getChildNodes element =
+    liftEffect do
+      nodes <- (Node.childNodes <<< Element.toNode) element
+      NodeList.toArray nodes
+
+  getElementText :: forall m. MonadEffect m => Element -> m String
+  getElementText = liftEffect <<< Node.textContent <<< Element.toNode
+
+  getAttribute :: forall m. MonadEffect m => MonadThrow ReadDomError m => String -> Element -> m String
+  getAttribute attr element = do
+    mValue <- liftEffect $ Element.getAttribute attr element
+    toMonadThrow $ note' (\_ -> MissingProperty element attr) mValue
```

Comment:
FYI, `Bolckly.Dom` is my favourite module ever 😄

Reply:
:)